crates/futures/src/nightly.rs

```diff
@@ -0,0 +1,252 @@
+        if let Poll::Ready(val) = self.resolved.poll_unpin(cx) {
+            return Poll::Ready(Ok(val.unwrap_throw()));
+        }
+
+        if let Poll::Ready(val) = self.rejected.poll_unpin(cx) {
+            return Poll::Ready(Err(val.unwrap_throw()));
+        }
+
+        Poll::Pending
+    }
+}
+
+/// Converts a Rust `Future` into a JavaScript `Promise`.
+///
+/// This function will take any future in Rust and schedule it to be executed,
+/// returning a JavaScript `Promise` which can then be passed back to JavaScript
+/// to get plumbed into the rest of a system.
+///
+/// The `future` provided must adhere to `'static` because it'll be scheduled
+/// to run in the background and cannot contain any stack references. The
+/// returned `Promise` will be resolved or rejected when the future completes,
+/// depending on whether it finishes with `Ok` or `Err`.
+///
+/// # Panics
+///
+/// Note that in wasm panics are currently translated to aborts, but "abort" in
+/// this case means that a JavaScript exception is thrown. The wasm module is
+/// still usable (likely erroneously) after Rust panics.
+///
+/// If the `future` provided panics then the returned `Promise` **will not
+/// resolve**. Instead it will be a leaked promise. This is an unfortunate
+/// limitation of wasm currently that's hoped to be fixed one day!
+pub fn future_to_promise<F>(future: F) -> Promise
+where
+    F: Future<Output = Result<JsValue, JsValue>> + 'static,
+{
+    let mut future = Some(future);
+
+    Promise::new(&mut |resolve, reject| {
+        // TODO change Promise::new to be FnOnce
+        spawn_local(future.take().unwrap_throw().map(move |val| {
+            match val {
+                Ok(val) => {
+                    resolve.call1(&JsValue::undefined(), &val).unwrap_throw();
+                },
+                Err(val) => {
+                    reject.call1(&JsValue::undefined(), &val).unwrap_throw();
+                },
+            }
+        }));
+    })
+}
+
+/// Runs a Rust `Future` on a local task queue.
+///
+/// The `future` provided must adhere to `'static` because it'll be scheduled
+/// to run in the background and cannot contain any stack references.
+///
+/// # Panics
+///
+/// This function has the same panic behavior as `future_to_promise`.
+pub fn spawn_local<F>(future: F)
+where
+    F: Future<Output = ()> + 'static,
+{
+    struct Task {
+        future: Mutex<Option<Pin<Box<dyn Future<Output = ()> + 'static>>>>,
+        is_queued: AtomicBool,
+    }
+
+    impl Task {
+        #[inline]
+        fn new<F>(future: F) -> Arc<Self> where F: Future<Output = ()> + 'static {
+            Arc::new(Self {
+                future: Mutex::new(Some(Box::pin(future))),
+                is_queued: AtomicBool::new(false),
+            })
+        }
+    }
+
+    impl ArcWake for Task {
+        fn wake_by_ref(arc_self: &Arc<Self>) {
+            // TODO can this be more relaxed ?
```

Comment:
Technically the ordering doesn't matter at all today in the sense that all orderings are equivalent when you're dealing with one thread of code. For multithreaded code wasm has no concept of orderings, so we'll want to stick `SeqCst` everywhere

Reply:
That's good to know about multi-threaded wasm.